### ISSUE TITLE
method to deal with default values

### ISSUE THREAD

**Issue by mephenor** [sbmlteam/jsbml#93]
Follow up on the discussions on libsbml-development to know what to do for default values when saving an SBML level 3 model.

Reported by: niko-rodrigue

**Comment by niko-rodrigue** [comment]
I think we did implement the initDefaults(int level, int version) methods but implementing setUnsetAttributesToL2v4Default might be good as well even now.

**Comment by draeger** [comment]
Personally, I would not support implementing a method with a prefix "setUnset" which is highly confusing. The `initDefaults(int level, int version)` should make clear what is meant. We should highlight this better in the documentation.

**Comment by niko-rodrigue** [comment]
I think the spirit of the setUnset method from libSBML is to initialize only the values that have not been set directly by the user. Like instead of calling initDefaults on all L3 elements at the start, you can call 'setUnsetXXX' (or whatever name but we might want to keep the libSBML name as well for compatibility) just before writing to be sure to have a valid model.
Also at the moment, the initDefaults method are not public on the SBase interface so you cannot call them easily, we have a pivotal tracker item about that (https://www.pivotaltracker.com/story/show/121664987).

**Comment by niko-rodrigue** [comment]
And the method hasRequiredAttributes would also be useful. We could, in general, re-review what utility methods libSBML provide to it's users around the default values and see which ones we can implement easily or want to implement.

**Comment by draeger** [comment]
Yes, so let's make `initDefaults(int level, int version)` public on `SBase` as suggested and also add this `hasAllRequiredAttributes` method, but not `setUnset` which sounds to me like "do undo" or "yes no" and also instead of having just one method with level and version as attributes we would have to implement n such specific methods for each Level/Version combination, which cannot be in our interest.

**Comment by niko-rodrigue** [comment]
in fact, checking the libsbml java api for Species (http://sbml.org/Software/libSBML/docs/java-api/org/sbml/libsbml/Species.html), it seems that they don't have any more the method setUnsetAttributesToL2v4Default, at least not publicly. And the only methods they have are `initDefaults()` and  `hasRequiredAttributes()` so that make it easier.

**Comment by dyrpsf** [comment]
Hi @draeger , @niko-rodrigue and @mephenor ,
I have opened a PR (#291 ) to address this issue! 

Based on the consensus in this thread, I implemented the following changes:
* Exposed `initDefaults(int level, int version)`, `initDefaults()`, and `hasRequiredAttributes()` as public methods directly on the `SBase` interface.
* Added the base implementations for these inside `AbstractSBase`. `hasRequiredAttributes()` now checks `isIdMandatory()`, and `initDefaults` is ready for subclasses to override with their specific defaults.
* Updated `SBMLDocument`, `CompartmentType`, and `SpeciesType` to make their existing `initDefaults()` methods `public` to satisfy the updated interface contract.

The core module compiles successfully and all tests are passing. Let me know if you'd like any adjustments!
